### ISSUE TITLE
Confused about cors implementation on API endpoints

### ISSUE THREAD

**Issue by lorenzosignoretti** [vercel/next.js#22274]
**What version of Next.js are you using?**

10.0.7

**What version of Node.js are you using?**

15.6.0

**What browser are you using?**

Chrome

**What operating system are you using?**

macOS

**How are you deploying your application?**

locally

**Describe the Bug**

The example provided in api-routes-cors suggests a way to use cors to restrict requests to /api/cors to POST / GET / OPTIONS methods only. Accordingly, using a different method from those mentioned should be blocked by CORS. This is not the case

An example of the same behavior has been documented in the discussion at this link: https://github.com/vercel/next.js/discussions/15395 and has been unanswered for months.

**Expected Behavior**

I would expect the cors implementation to be able to throw an error when the parameters do not conform to the provided options (e.g. wrong method, not allowed origin, etc).

**To Reproduce**

Clone the example provided by Vercel at https://github.com/vercel/next.js/blob/canary/examples/api-routes-cors/. Run locally and try making a DELETE request to localhost:3000/api/cors. The response has a 200 status.

Hi team,

As detailed above, I am confused by Next's suggested cors implementation. Regardless of the options provided, all requests to an API endpoint continue to be allowed.

How can I – for example restrict – an API endpoint to a few allowed origins only?

Thank you in advance for the help.

**Comment by lorenzosignoretti** [comment]
I am aware [CORS is a browser security feature](https://developer.mozilla.org/nl/docs/Web/HTTP/CORS). However, it is indeed possible to block an incoming request from a non-whitelisted url through it.

In case anyone else was looking into implementing the same type of behavior (like https://github.com/vercel/next.js/issues/20470), the gist of the solution is to validate the origin dynamically using a function provided to the origin option of cors. We can use that to throw an error that can be intercepted inside a try-catch block.

The pseudo-code of the suggested solution would look something like:
 ```javascript
var whitelist = ['http://example1.com', 'http://example2.com'];

const cors = initMiddleware(
  // You can read more about the available options here: https://github.com/expressjs/cors#configuration-options
  Cors({
    // Only allow requests with GET, POST and OPTIONS
    methods: ['GET', 'POST', 'OPTIONS'],
    origin: function (origin, callback) {
      if (whitelist.indexOf(origin) !== -1) {
        callback(null, true);
      } else {
        callback(new Error('Not allowed by CORS'));
      }
    },
  })
);

export default async function handler(req, res) {
  try{
    await cors(req, res);
    // rest of your code implementation
  } catch (error) {
    return res.status(500).send(error.message)
  }
}
```

**Comment by benjaminviaud** [comment]
Hello @lorenzosignoretti,

Thanks for your example.

I try to allowed only GET methods for example but it has no effect. Do you have a suggestion?

Thanks.

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
